profiling-ffi/src/exporter.rs

```diff
@@ -165,128 +169,125 @@ impl From<RequestBuildResult> for Result<Box<Request>, String> {
+                Ok(request) => {
+                    RequestBuildResult::Ok(NonNull::new_unchecked(Box::into_raw(Box::new(request))))
+                }
                 Err(err) => RequestBuildResult::Err(err.into()),
             }
         }
     }
 }
 
 /// # Safety
-/// The `request` must be valid, or null. Notably, it is invalid if it's been
-/// sent to `ddog_prof_Exporter_send` or `ddog_prof_Exporter_Request_drop`
-/// already.
+/// The `request` must be valid, or null. After this call, the `request` will
+/// be invalid.
 #[no_mangle]
 pub unsafe extern "C" fn ddog_prof_Exporter_Request_drop(request: *mut Request) {
     if !request.is_null() {
-        drop(Box::from_raw(request));
+        drop_in_place(request)
     }
 }
 
 /// Sends the request, returning the HttpStatus.
 ///
 /// # Arguments
 /// * `exporter` - Borrows the exporter for sending the request.
-/// * `request` - Takes ownership of the request. Do not call
-///               `ddog_prof_Exporter_Request_drop` on `request` after calling
-///               `ddog_prof_Exporter_send` on it.
+/// * `request` - Takes ownership of the request, replacing it with a null
+///               pointer. This is why it takes a double-pointer, rather than
+///               a single one.
 /// * `cancel` - Borrows the cancel, if any.
 ///
 /// # Safety
 /// All non-null arguments MUST have been created by created by apis in this module.
 #[no_mangle]
 #[must_use]
 pub unsafe extern "C" fn ddog_prof_Exporter_send(
-    exporter: Option<NonNull<ProfileExporter>>,
-    request: *mut Request,
-    cancel: Option<NonNull<CancellationToken>>,
+    exporter: *mut ProfileExporter,
+    request: *mut *mut Request,
+    cancel: *mut CancellationToken,
 ) -> SendResult {
+    match ddog_prof_exporter_send_impl(exporter, request, cancel) {
+        Ok(code) => SendResult::HttpResponse(code),
+        Err(err) => SendResult::Err(Error::from(err.context("failed ddog_prof_Exporter_send"))),
+    }
+}
+
+unsafe fn ddog_prof_exporter_send_impl(
+    exporter: *mut ProfileExporter,
+    request: *mut *mut Request,
+    cancel: *mut CancellationToken,
+) -> anyhow::Result<HttpStatus> {
     // Re-box the request first, to avoid leaks on other errors.
-    let request = if request.is_null() {
-        return SendResult::Err(Error::from("failed to export: request was null"));
-    } else {
-        Box::from_raw(request)
+    let request_ptr = match request.as_mut() {
+        Some(r) => r,
+        None => anyhow::bail!("request was null"),
     };
-
-    let exp_ptr = match exporter {
-        None => {
-            return SendResult::Err(Error::from("failed to export: exporter was null"));
-        }
-        Some(e) => e,
+    let mut tmp = std::ptr::null_mut(); // leave a nullptr for the caller
+    std::mem::swap(&mut tmp, request_ptr);
+    let request = match tmp.as_mut() {
+        Some(r) => Box::from_raw(r as *mut _),
+        None => anyhow::bail!("request was null"),
     };
```

Comment:
This seems off -- do we really need to check if the request is null twice?

Reply:
It's `** Request`,  so there are two null checks, one for each pointer.